File: lib/features/transactions/domain/models/expense.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:savvy/core/constants/financial_enums.dart';

part 'expense.freezed.dart';
part 'expense.g.dart';

@freezed
abstract class Expense with _$Expense {
  const factory Expense({
    required String id,
    required double amount,
    required ExpenseCategory category,
    @Default(ExpenseType.variable) ExpenseType expenseType,
    String? subcategory,
    String? person,
    required DateTime date,
    String? note,
    @Default(false) bool isRecurring,
    DateTime? recurringEndDate,
    @Default(false) bool isDeleted,
    required DateTime createdAt,
    /// Per-month amount overrides for recurring items.
    /// Key: "YYYY-MM", Value: override amount for that month.
    /// Months not present use the default [amount].
    @Default({}) Map<String, double> monthlyOverrides,
    /// true = ödendi (gider ödendi), false = beklemede
    @Default(false) bool isSettled,
  }) = _Expense;

  factory Expense.fromJson(Map<String, dynamic> json) =>
      _$ExpenseFromJson(json);
}

File: SavvyApp/Packages/Core/SavvyFoundation/Sources/SavvyFoundation/Models/Expense.swift
import Foundation

public struct Expense: FirestoreEntity {
    public let id: String
    public var amount: Decimal
    public var category: ExpenseCategory
    public var expenseType: ExpenseType
    public var subcategory: String?
    public var person: String?
    public var date: Date
    public var note: String?
    public var isRecurring: Bool
    public var recurringEndDate: Date?
    public var monthlyOverrides: [String: Decimal]
    public var isDeleted: Bool
    public let createdAt: Date

    public init(
        id: String = UUID().uuidString,
        amount: Decimal,
        category: ExpenseCategory,
        expenseType: ExpenseType = .variable,
        subcategory: String? = nil,
        person: String? = nil,
        date: Date = Date(),
        note: String? = nil,
        isRecurring: Bool = false,
        recurringEndDate: Date? = nil,
        monthlyOverrides: [String: Decimal] = [:],
        isDeleted: Bool = false,
        createdAt: Date = Date()
    ) {
        self.id = id
        self.amount = amount
        self.category = category
        self.expenseType = expenseType
        self.subcategory = subcategory
        self.person = person
        self.date = date
        self.note = note
        self.isRecurring = isRecurring
        self.recurringEndDate = recurringEndDate
        self.monthlyOverrides = monthlyOverrides
        self.isDeleted = isDeleted
        self.createdAt = createdAt
    }
}
